src/gennylib/include/gennylib/topology.hpp

```diff
@@ -0,0 +1,219 @@
+// Copyright 2021-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#ifndef HEADER_6D7AD2AF_7CD1_49E4_9712_741ABD354DFC_INCLUDED
+#define HEADER_6D7AD2AF_7CD1_49E4_9712_741ABD354DFC_INCLUDED
+
+#include <vector>
+#include <variant>
+#include <mutex>
+#include <sstream>
+
+#include <gennylib/v1/PoolFactory.hpp>
+#include <gennylib/service.hpp>
+
+namespace genny {
+
+class AbstractTopologyDescription;
+class MongodDescription;
+class MongosDescription;
+class ReplSetDescription;
+class ShardedDescription;
+
+/**
+ * Inherit from TopologyVisitor and override particular visit methods
+ * to implement an algorithm that operates on a cluster. Pass the visitor
+ * to a Topology object to execute.
+ *
+ * The idea is to create a visitor that focuses on each part of the cluster,
+ * allow the Topology object to handle traversal, and keep application-level
+ * code as topology-agnostic as possible.
+ */
+class TopologyVisitor {
+public:
+    virtual void visitTopLevelPre(const AbstractTopologyDescription&) {}
+    virtual void visitTopLevelPost(const AbstractTopologyDescription&) {}
+
+    virtual void visitMongodDescription(const MongodDescription&) {}
+    virtual void visitMongosDescription(const MongosDescription&) {}
+
+    virtual void visitReplSetDescriptionPre(const ReplSetDescription&) {}
+    virtual void visitReplSetDescriptionPost(const ReplSetDescription&) {}
+
+    virtual void visitShardedDescriptionPre(const ShardedDescription&) {}
+    virtual void visitShardedDescriptionPost(const ShardedDescription&) {}
+
+
+    /** Misc hooks that most visitors won't need. **/
+
+    // Called between mongods in a replica set.
+    virtual void visitMongodDescriptionBetween(const ReplSetDescription&) {}
+
+    // Called before/after/between visiting shards.
+    virtual void visitShardsPre(const ShardedDescription&) {}
+    virtual void visitShardsPost(const ShardedDescription&) {}
+    virtual void visitShardsBetween(const ShardedDescription&) {}
+
+    // Called before/after/between visiting mongoses.
+    virtual void visitMongosesPre(const ShardedDescription&) {}
+    virtual void visitMongosesPost(const ShardedDescription&) {}
+    virtual void visitMongosesBetween(const ShardedDescription&) {}
```

Comment:
1. Why not pure-virtual (`= 0`)?
2. If most visitors won't need them maybe just don't include them for now?

Reply:
1. The visit methods can't be pure-virtual because they're always called in the topology description traversal, even as a nop. The idea is that subclass visitors only need to override the visit methods they care about, instead of overriding to add a bunch of nops.
2. The `JsonVisitor` needs them. That visitor isn't 100% necessary but it was useful for testing both the discovery and that traversal works correctly.